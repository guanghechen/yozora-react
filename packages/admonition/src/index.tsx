import cn from 'clsx'
import PropTypes from 'prop-types'
import React from 'react'

export interface AdmonitionProps {
  /**
   * Admonition keyword.
   */
  keyword?: string
  /**
   * Admonition title
   */
  title?: React.ReactNode
  /**
   * Admonition contents.
   */
  children?: React.ReactNode
  /**
   * Root css class of the component.
   */
  className?: string
  /**
   * Root css style.
   */
  style?: React.CSSProperties
}

/**
 * Render yozora `admonition`.
 *
 * @see https://www.npmjs.com/package/@yozora/ast#admonition
 * @see https://www.npmjs.com/package/@yozora/tokenizer-admonition
 */
export function Admonition(props: AdmonitionProps): React.ReactElement {
  const { className, style, keyword = 'default', title, children } = props

  let modifier = keyword.trim().toLowerCase()
  switch (modifier) {
    case 'note':
      modifier = 'default'
      break
    case 'important':
      modifier = 'info'
      break
    case 'tip':
      modifier = 'success'
      break
    case 'caution':
      modifier = 'warning'
      break
    case 'error':
      modifier = 'danger'
      break
  }

  return (
    <div
      className={cn(
        'yozora-admonition',
        'yozora-admonition--' + modifier,
        className,
      )}
      style={style}
    >
      <div key="heading" className="yozora-admonition__heading">
        <h5>{title}</h5>
      </div>
      <div key="body" className="yozora-admonition__body">
        {children}
      </div>
    </div>
  )
}

Admonition.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  keyword: PropTypes.string,
  style: PropTypes.object,
  title: PropTypes.node,
}

Admonition.displayName = 'YozoraAdmonition'
export default Admonition
