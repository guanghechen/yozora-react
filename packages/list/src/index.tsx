export * from './item'
export * from './olist'
export * from './ulist'
