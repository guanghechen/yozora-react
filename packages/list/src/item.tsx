import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'


/**
 * Props for creating ListItem
 */
export interface ListItemProps extends React.LiHTMLAttributes<HTMLSpanElement> {
  /**
   * Whether if is a TODO item, and given its status
   */
  status?: 'todo' | 'doing' | 'done'
  /**
   * ListItem content
   */
  children: React.ReactNode
}


const Container = styled.li`
  margin: var(--md-list-item-margin, 0 0 0 .25em);
`


/**
 * Render ListItem content
 *
 * @param props
 */
export const ListItem = React.forwardRef<HTMLLIElement, ListItemProps>(
  (props, forwardRef): React.ReactElement => {
    const { status, ...listItemProps } = props
    return (
      <Container { ...listItemProps } ref={ forwardRef } />
    )
  }
)


ListItem.propTypes = {
  status: PropTypes.oneOf<'todo' | 'doing' | 'done'>(['todo', 'doing', 'done']),
  children: PropTypes.node.isRequired,
}


ListItem.displayName = 'ListItem'


export default ListItem
