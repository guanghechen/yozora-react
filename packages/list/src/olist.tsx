import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'


/**
 * Props for ordered list
 */
export interface OrderedListProps extends React.OlHTMLAttributes<HTMLOListElement> {
  /**
   * Start number
   */
  start: number
  /**
   * Ordered list content
   */
  children: React.ReactNode
}


const Container = styled.ol`
  padding: var(--md-list-padding, 0 0 0 1.2em);
  margin: var(--md-list-margin, 0 0 1em);
`


/**
 * Ordered list
 * @param props
 */
export const OrderedList = React.forwardRef<HTMLOListElement, OrderedListProps>(
  (props, forwardRef): React.ReactElement => {
    const { start, children, ...olProps } = props
    return (
      <Container { ...olProps } start={ start } ref={ forwardRef }>
        { children }
      </Container>
    )
  }
)


OrderedList.propTypes = {
  start: PropTypes.number.isRequired,
  children: PropTypes.node.isRequired,
}


OrderedList.displayName = 'OrderedList'


export default OrderedList
