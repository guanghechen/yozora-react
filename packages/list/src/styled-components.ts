import 'styled-components'
import type { YozoraListTheme } from './theme'


declare module 'styled-components' {
  export interface DefaultTheme {
    yozora: YozoraTheme
  }

  interface YozoraTheme {
    list?: YozoraListTheme
  }
}
