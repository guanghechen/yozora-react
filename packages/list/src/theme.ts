import type { CSSProperties, DefaultTheme } from 'styled-components'

/**
 * yozora `list` theme
 */
export interface YozoraListTheme {
  /**
   * List padding
   * @default '0 0 0 1.2em'
   */
  padding?: CSSProperties['padding'] | string
  /**
   * List margin
   * @default '0 0 1em'
   */
  margin?: CSSProperties['margin'] | string
  /**
   * List line height
   * @default 2
   */
  lineHeight?: CSSProperties['lineHeight'] | string
  /**
   * List font color
   * @default 'inherit'
   */
  color?: CSSProperties['color'] | string
}

/**
 * Default yozora `list` theme
 */
export const defaultListTheme: YozoraListTheme = {
  padding: '0 0 0 1.2em',
  margin: '0 0 1em',
  lineHeight: 2,
  color: 'inherit',
}

/**
 * Get `list` style
 * @param key
 * @param defaultTheme
 */
export function getListStyle(
  key: keyof YozoraListTheme,
  defaultTheme: YozoraListTheme = defaultListTheme,
): (props: { theme: DefaultTheme }) => YozoraListTheme[typeof key] {
  return props => {
    const { yozora } = props.theme
    if (yozora == null || yozora.list == null || yozora.list[key] == null)
      return defaultTheme[key]
    return yozora.list[key]
  }
}
