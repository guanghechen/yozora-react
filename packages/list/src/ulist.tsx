import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'


/**
 * Props for unordered list
 */
export interface UnorderedListProps extends React.HTMLAttributes<HTMLUListElement> {
  /**
   *
   */
  children: React.ReactNode
}


const Container = styled.ul`
  padding: var(--md-list-padding, 0 0 0 1.2em);
  margin: var(--md-list-margin, 0 0 1em);
`

/**
 * Ordered list
 * @param props
 */
export const UnorderedList = React.forwardRef<HTMLUListElement, UnorderedListProps>(
  (props, forwardRef): React.ReactElement => {
    const { children, ...olProps } = props
    return (
      <Container { ...olProps } ref={ forwardRef }>
        { children }
      </Container>
    )
  }
)


UnorderedList.propTypes = {
  children: PropTypes.node.isRequired,
}


UnorderedList.displayName = 'UnorderedList'


export default UnorderedList
