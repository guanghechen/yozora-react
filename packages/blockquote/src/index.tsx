import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import './styled-components'
import { defaultBlockquoteTheme, getBlockquoteStyle } from './theme'
export * from './theme'


/**
 * Props for creating Blockquote
 */
export interface BlockquoteProps extends React.BlockquoteHTMLAttributes<HTMLDivElement> {
  /**
   * Blockquote content
   */
  children: React.ReactNode
}


/**
 * Render Blockquote content
 *
 * @param props
 */
export const Blockquote = React.forwardRef<HTMLDivElement, BlockquoteProps>(
  (props, forwardRef): React.ReactElement => {
    return (
      <Container { ...props } ref={ forwardRef } />
    )
  }
)


Blockquote.propTypes = {
  children: PropTypes.node.isRequired,
}


Blockquote.displayName = 'YozoraBlockquote'
export default Blockquote


const Container = styled.blockquote`
  padding: ${ getBlockquoteStyle('padding') };
  border-left: 0.25em solid ${ getBlockquoteStyle('borderColor') };
  margin: ${ getBlockquoteStyle('margin') };
  background:  ${ getBlockquoteStyle('background') };
  color: ${ getBlockquoteStyle('color') };
`


Container.defaultProps = {
  theme: { yozora: { blockquote: defaultBlockquoteTheme } }
}


export const BlockquoteClasses = {
  container: `${ Container }`,
}
