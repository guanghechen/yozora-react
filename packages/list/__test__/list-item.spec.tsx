import React from 'react'
import { mount, shallow } from 'enzyme'
import { ListItem } from '../src'


describe('basic rendering case', () => {
  const errorLogger = jest
    .spyOn(global.console, 'error')
    .mockImplementation((...args) => {
      throw new Error(args.join(' '))
    })

  afterAll(() => {
    errorLogger.mockRestore()
  })

  it('render a simple content', () => {
    const text = 'Hello, world!'
    const wrapper = shallow(
      <ListItem key={ 1 }>
        <span>
          <ListItem key={ 1 }>{ text }</ListItem>
        </span>
      </ListItem>
    )
    expect(wrapper.text()).toEqual(text)
  })

  it('render with custom className', () => {
    const text = 'Hello, world!'
    const className = 'custom-list-item'
    const wrapper = shallow(
      <ListItem key={ 1 } className={ className }>
        <span>{ text }</span>
      </ListItem>
    )
    expect(wrapper.hasClass(className)).toEqual(true)
    expect(wrapper.text()).toEqual(text)
  })

  it('children is required', () => {
    for (const value of [undefined, null] as any[]) {
      expect(() => {
        mount(
          <ListItem key={ 1 }>
            { value }
          </ListItem>
        )
      }).toThrow(/Failed prop type: The prop `children` is marked as required/)
    }
  })

  it('forward ref', () => {
    const ref = React.createRef<HTMLLIElement>()
    const wrapper = mount(
      <ListItem ref={ ref } key={ 1 } data-value="waw">
        1
      </ListItem>
    )

    const o = wrapper.getDOMNode()
    expect(o).toEqual(ref.current)
    expect(o.getAttribute('data-value')).toEqual('waw')
  })

  it('snapshot', () => {
    const wrapper = mount(
      <ListItem key={ 1 } status="done" style={ { color: 'orange', fontSize: '16px' } }>
        some text1
        <span>some text2</span>
      </ListItem>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
