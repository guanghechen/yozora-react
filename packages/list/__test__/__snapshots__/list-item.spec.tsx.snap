// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic rendering case snapshot 1`] = `
.c0 {
  margin: var(--md-list-item-margin,0 0 0 .25em);
}

<ListItem
  status="done"
  style={
    Object {
      "color": "orange",
      "fontSize": "16px",
    }
  }
>
  <styled.li
    style={
      Object {
        "color": "orange",
        "fontSize": "16px",
      }
    }
  >
    <li
      className="c0"
      style={
        Object {
          "color": "orange",
          "fontSize": "16px",
        }
      }
    >
      some text1
      <span>
        some text2
      </span>
    </li>
  </styled.li>
</ListItem>
`;
